tiledb/sm/query/query_condition.cc

```diff
@@ -1078,34 +1069,94 @@ Status QueryCondition::apply_clause_dense(
-    const uint64_t length,
     const uint64_t src_cell,
     const uint64_t stride,
-    uint8_t* result_buffer) {
-  // Iterate through each clause.
-  // This assumes all clauses are combined with a logical "AND".
-  for (const auto& clause : clauses_) {
-    RETURN_NOT_OK(apply_clause_dense(
-        clause,
+    std::vector<uint8_t>& result_buffer) const {
+  if (node->get_tag() == ASTNodeTag::VAL) {
+    const ASTNodeVal* node_ptr = dynamic_cast<ASTNodeVal*>(node.get());
+    Status s = apply_val_dense(
+        node_ptr,
         array_schema,
         result_tile,
         start,
-        length,
         src_cell,
         stride,
-        result_buffer));
+        result_buffer);
+    if (!s.ok()) {
+      throw std::runtime_error(
+          "From apply_tree_dense: apply_val_dense failed.");
+    }
+  } else {
+    // for each child, declare a new result buffer, combine it all into the
+    // other one based on combination op
+    auto expr_ptr = dynamic_cast<ASTNodeExpr*>(node.get());
+    if (expr_ptr->combination_op_ == QueryConditionCombinationOp::AND) {
+      std::fill(result_buffer.begin(), result_buffer.end(), 1);
+    } else if (expr_ptr->combination_op_ == QueryConditionCombinationOp::OR) {
+      std::fill(result_buffer.begin(), result_buffer.end(), 0);
+    }
+    for (const auto& child : expr_ptr->nodes_) {
+      std::vector<uint8_t> child_result_buffer(result_buffer.size(), 1);
+      apply_tree_dense(
+          child,
+          array_schema,
+          result_tile,
+          start,
+          src_cell,
+          stride,
+          child_result_buffer);
+
+      switch (expr_ptr->combination_op_) {
+        case QueryConditionCombinationOp::AND: {
+          for (uint64_t c = 0; c < result_buffer.size(); ++c) {
+            result_buffer[c] &= child_result_buffer[c];
+          }
+        } break;
+        case QueryConditionCombinationOp::OR: {
+          for (uint64_t c = 0; c < result_buffer.size(); ++c) {
+            result_buffer[c] |= child_result_buffer[c];
+          }
+        } break;
+        case QueryConditionCombinationOp::NOT: {
+          throw std::runtime_error(
+              "QueryCondition NOT operator not supported in TileDB system.");
+        } break;
+        default: {
+          throw std::logic_error(
+              "QueryCondition operator apply_tree_dense case, should not get "
+              "here.");
+        }
+      }
+    }
   }
+}
 
+Status QueryCondition::apply_dense(
+    const ArraySchema& array_schema,
+    ResultTile* result_tile,
+    const uint64_t start,
+    const uint64_t length,
+    const uint64_t src_cell,
+    const uint64_t stride,
+    uint8_t* result_buffer) {
+  // Iterate through the tree.
+  std::vector<uint8_t> result_vector(length, 1);
+  apply_tree_dense(
+      tree_, array_schema, result_tile, start, src_cell, stride, result_vector);
+
+  // copy into buffer
+  std::copy(result_vector.begin(), result_vector.end(), result_buffer + start);
```

Comment:
I was told that I should try to make the code support more C++ style operations, which is why I pass in vectors and do a copy at the end.

Reply:
This will cause a major performance impact, we absolutely need to prevent copying data whenever possible. There might be some structures we can use to make to make this look more like c++ (span, etc) that have a high chance of getting optimized out at compilation, but the extra vector will generate an extra copy. In the dense case, you could be doing this billion of times.

Reply:
Solved using new design for apply